### PR TITLE
Add Spring Boot dependencies

### PR DESCRIPTION
**Pull request by cuijulian**
Jira: https://issues.jboss.org/browse/PLANNER-1695

Prerequisite for:

- https://github.com/kiegroup/optaweb-employee-rostering/pull/312
- https://github.com/kiegroup/optaweb-vehicle-routing/pull/211
- https://github.com/kiegroup/droolsjbpm-integration/pull/1931
- https://github.com/kiegroup/jbpm-work-items/pull/114

### CODE REVIEW

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
are you also planning to remove them in our kiegroup repositories to not duplicate them?

**Comment by cuijulian** [comment]
> are you also planning to remove them in our kiegroup repositories to not duplicate them?

Yes, I will remove the Spring Boot dependencies from the repositories that use `kie-parent` as the parent POM (which to my knowledge are only `optaweb-vehicle-routing` and `optaweb-employee-rostering`). This is to ensure the same version is used.

**Comment by mareknovotny** [comment]
good, there is also https://github.com/kiegroup/droolsjbpm-integration/tree/master/kie-spring-boot referencing and using it

**Comment by cuijulian** [comment]
> good, there is also https://github.com/kiegroup/droolsjbpm-integration/tree/master/kie-spring-boot referencing and using it

Good to know, will also send a PR there.

**Comment by ge0ffrey** [comment]
@mareknovotny Can we get this merged?

@cuijulian In a seperate PR, we'll want to upgrade <version.org.springframework> to 5.2.0, but this might affect droolsjbpm-integration.

**Comment by mareknovotny** [comment]
@ge0ffrey are we talking about Spring boot version? I guess a typo in your comment  *5.2.0*

Anyway the dependent PRs are not here, one in draft is failing.

**Comment by cuijulian** [comment]
> @ge0ffrey are we talking about Spring boot version? I guess a typo in your comment _5.2.0_
> 
> Anyway the dependent PRs are not here, one in draft is failing.

@mareknovotny Yes, we're talking about the Spring Boot version (at least for this PR). I think Geoffrey was making a suggestion for a future PR. I'm having some build issues with some of the projects, I will send PR's when they're resolved.

**Comment by MarianMacik** [comment]
FYI this means we are upgrading from 2.1.1 to 2.2.0 in the droolsjbpm-integration repo.

**Comment by ge0ffrey** [comment]
@mareknovotny Note that spring-core is currently on 4.x. That doesn't work with spring-boot 2.1.x (what we have now?) nor 2.2.x (what this PR will change it into). So no regression :)
But, in a 2th PR, we need to get spring-core on 5.2, so it's compatible with the spring-boot version.

**Comment by cuijulian** [comment]
> @cuijulian from the engine side it is only in droolsjbpm-integration. AFAIK that is the only place where we use Spring/SpringBoot other than OptaWeb demos.

@MarianMacik Thanks for confirming, that's what I found too.

**Comment by mareknovotny** [comment]
so can we get springframework upgraded here to 5.x too, as if it doesn't work with springboot 2.2.0 we are breaking the build by merging this, am I correct?

**Comment by ge0ffrey** [comment]
Jenkins execute full downstream build

**Comment by ge0ffrey** [comment]
@mswiderski Already upgraded kie-spring to spring 5.1.x, so maybe upgrading kie-parent from spring 4 to spring 5.2 won't be a big pain. @mareknovotny see
https://github.com/kiegroup/droolsjbpm-integration/blame/master/kie-spring/pom.xml#L23

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by mswiderski** [comment]
@ge0ffrey fine with me, just let's make sure that Fuse/RHOAR is fine with moving to spring boot 2.2 so we don't diverge again. Other than that it's fine for me to upgrade

**Comment by mbiarnes** [comment]
@cuijulian can you please resolve this conflicts?

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
@cuijulian see enforcer ban duplicated classes on classpath now:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) on project kie-soup-dataset-elasticsearch: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-duplicated-classes) on project kie-soup-dataset-elasticsearch: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.
```

File: pom.xml [hunk 1]
```diff
@@ -439,11 +440,10 @@
     <!-- Make OSGi happy -->
     <osgi.snapshot.qualifier>${maven.build.timestamp}</osgi.snapshot.qualifier>
 
-    <version.io.netty>4.1.42.Final</version.io.netty>
     <!-- IMPORTANT: Don't use this dependency. The right dependency version is the one named as "version.io.netty".
          This is just needed by Elasticsearch Transport Plugin because it has a compatibility mode with netty 3 and
          it can't be removed -->
-    <version.io.netty.old>3.10.6.Final</version.io.netty.old>
+    <version.io.netty>3.10.6.Final</version.io.netty>
```
Comment (MarianMacik) [inline pom.xml]: Why are we basically reverting the change from an already merged PR? https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1095/files#diff-600376dffeb79835ede4a0b285078036R446

Comment (mareknovotny) [reply pom.xml]: good catch @MarianMacik this should not be reverted.

Is it a matter of rebasing or intentionally changed @cuijulian ?

Comment (cuijulian) [reply pom.xml]: Sorry about this, yes it is from not rebasing on `master`. I will force push changes once the issues are resolved.

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
> There are karaf failures in droolsjbpm-integration repository

@mareknovotny The duplicate classes in karaf were present before this PR afaik. It is the `kie-wb-common` repo that is breaking the downstream build, I will look into it.

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
@mareknovotny @MarianMacik There is an issue with this PR that I realized (thank you @yurloc for bringing this to my attention). Currently it imports the `spring-boot-dependencies` BOM into `kie-parent`, which would override the dependency management of KIE (which explains the many broken builds in the full downstream build). I see two ways to resolve this:

1. Remove the Spring Boot BOM from `kie-parent` and manually add all the Spring Boot dependencies required by the KIE team into the `<dependencyManagement>` section of the parent.
2. Put `spring-boot-dependencies` BOM first in `kie-parent` so the rest of the KIE imports override it.

Are there any other options that you think I'm missing? What would be the best option for the team?

**Comment by mareknovotny** [comment]
@cuijulian I am not sure I understand your mentioned options and reasons. As you have not referenced a jira or description of problem what are you trying to solve here first. I can only guess that you are trying to move individual dep mgmts from optaweb examples and droolsjbpm-integration to one place up, is that right?
Then you will need import only spring-boot BOM for managing of deps and versions. If you need to change/override some of the deps in the spring boot BOM you need to move individual dep overrides before the spring boot BOM as there is the order what matters - first declaration is taken into mind by Maven.

**Comment by cuijulian** [comment]
@mareknovotny I added a link to the Jira in the PR comment, sorry for the delay. I am trying to bump the `springframework` version and add `springframework.boot` version management so that all projects managed by `kie-parent` use the same version of Spring. I've decided against importing the Spring Boot BOM to `kie-parent` because it overrides KIE managed dependencies (the issue I described above). I will update the PR to reflect this.

**Comment by mareknovotny** [comment]
I looked what actually https://search.maven.org/artifact/org.springframework.boot/spring-boot-dependencies/2.2.1.RELEASE/pom contains and to import BOM for dependency management is ok if it manages only these spring boot deps, not the 3rd party deps which spring boot needs.

That would not be accepted here and you already realized there are additional 3 rd party deps which can colide with our own required 3rd party deps.

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
@MarianMacik Sorry to bother you again. The full downstream build is unstable due to this [failed test](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/droolsjbpm-build-bootstrap-downstream-pullrequests/218/testReport/junit/org.kie.server.client/LoadBalancerClientTest/testDefaultLoadBalancerNoServersAvailable/). However, when I run `mvn clean install -Dfull` in the `kie-server-client` module I cannot reproduce the test failure on my local machine. I will try on another machine. If you get a chance could you try reproducing that test failure? I'm not sure what's causing the discrepancy between Jenkins and my machine. Thanks in advance.

**Comment by mareknovotny** [comment]
Jenkins execute full downstream build

**Comment by cuijulian** [comment]
Jenkins execute full downstream build

**Comment by MarianMacik** [comment]
I think that because it works on the droolsjbpm-integration itself, it shouldn't be an issue in FDB. Locally it works for me as well with droolsjbpm-buill-bootstrap and kie-server-parent built from the PRs. I think we should try one more time.


Jenkins execute full downstream build

**Comment by cuijulian** [comment]
@MarianMacik Build passed!! Thanks for your help! Is there anything I need to do before merge?

**Comment by MarianMacik** [comment]
@mareknovotny can you approve as well?

**Comment by mareknovotny** [comment]
ok, it is good to see green, what about the dependent PRs are they green all too?

**Comment by cuijulian** [comment]
@mareknovotny Some dependent PRs fail because they do not compile upstream, but they will work if this PR is merged. I thought if the full downstream build passes then that means the dependent PR's passed as well, is that true? In any case we should wait for approval for the dependent PR's, thank you for requesting reviews for those.

**Comment by mareknovotny** [comment]
I was not sure if the FDB was run with all the latest changes, but as you wrote above, if FDB is green all should be ok

**Comment by MarianMacik** [comment]
Yes, the FDB is fresh with all changes in it.

**Comment by mareknovotny** [comment]
so we need now all approvals in downstream repositories

File: pom.xml [hunk 2]
```diff
@@ -5145,13 +5146,14 @@
...
         <version>${version.org.springframework}</version>
       </dependency>
+
       <dependency>
-        <groupId>org.springframework</groupId>
-        <artifactId>spring-websocket</artifactId>
-        <version>${version.org.springframework}</version>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-dependencies</artifactId>
+        <version>${version.org.springframework.boot}</version>
```
Comment (yurloc) [inline pom.xml]: Let's remove this whole dependency. I've checked and I couldn't import the BOM without specifying version so this actually doesn't work as I expected.

Comment (yurloc) [reply pom.xml]: But we can do that in another PR. This has been open for too long.

Comment (cuijulian) [reply pom.xml]: Noted. I'll make a Jira for this.

**Comment by cuijulian** [comment]
@mareknovotny All dependent PR's have approval. Is there anything else I need to do before merge?

**Comment by ge0ffrey** [comment]
If all PR's are green then it's ready to merge, so starting the merge.

**Comment by ge0ffrey** [comment]
Merging this group PR in it's entirety.

**Pull request merged**